File: src-tauri/crates/bluey-core/src/types/session.rs
use serde::{Deserialize, Serialize};

use super::response::BlueyResponse;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SessionStatus {
    Active,
    Paused,
    Completed,
}

/// Mirrors `Session`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Session {
    pub id: String,
    pub mode_id: String,
    pub started_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ended_at: Option<String>,
    pub status: SessionStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Map<String, serde_json::Value>>,
}

/// Mirrors `SessionEventType`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SessionEventType {
    SessionStarted,
    SessionPaused,
    SessionResumed,
    SessionEnded,
    QuestionDetected,
    CodingProblemDetected,
    ObjectionDetected,
    DecisionDetected,
    ActionItemDetected,
    TopicChange,
    ImportantStatement,
    FollowUp,
    ScreenCaptured,
    ResponsePrepared,
    ResponseGenerated,
    ResponseFailed,
    DocumentAttached,
    ModeChanged,
    NoteAdded,
    SummaryGenerated,
}

/// Mirrors `SessionEvent`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionEvent {
    pub id: String,
    pub session_id: String,
    #[serde(rename = "type")]
    pub event_type: SessionEventType,
    pub title: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub refs: Option<std::collections::BTreeMap<String, String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub confidence: Option<f32>,
    pub created_at: String,
}

/// Mirrors `SessionNote`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionNote {
    pub id: String,
    pub session_id: String,
    pub content: String,
    pub created_at: String,
    pub updated_at: String,
}

/// Mirrors `SessionSummary`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SessionSummary {
    pub id: String,
    pub session_id: String,
    pub mode_id: String,
    pub overview: String,
    #[serde(default)]
    pub topics: Vec<String>,
    #[serde(default)]
    pub questions: Vec<String>,
    #[serde(default)]
    pub answers: Vec<String>,
    #[serde(default)]
    pub decisions: Vec<String>,
    #[serde(default)]
    pub action_items: Vec<String>,
    #[serde(default)]
    pub open_items: Vec<String>,
    #[serde(default)]
    pub improvements: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sections: Option<Vec<SummarySection>>,
    pub created_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SummarySection {
    pub title: String,
    pub content: String,
}

/// Input for `sessions_save_summary` (summary without id/createdAt).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionSummaryInput {
    pub session_id: String,
    pub mode_id: String,
    pub overview: String,
    #[serde(default)]
    pub topics: Vec<String>,
    #[serde(default)]
    pub questions: Vec<String>,
    #[serde(default)]
    pub answers: Vec<String>,
    #[serde(default)]
    pub decisions: Vec<String>,
    #[serde(default)]
    pub action_items: Vec<String>,
    #[serde(default)]
    pub open_items: Vec<String>,
    #[serde(default)]
    pub improvements: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sections: Option<Vec<SummarySection>>,
}

/// Mirrors `SessionListItem`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionListItem {
    pub session: Session,
    pub mode_name: String,
    pub event_count: u32,
    pub response_count: u32,
    pub transcript_segment_count: u32,
    pub has_summary: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub snippet: Option<String>,
}

/// Mirrors `SessionDetail`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionDetail {
    pub session: Session,
    pub events: Vec<SessionEvent>,
    pub notes: Vec<SessionNote>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub summary: Option<SessionSummary>,
    pub responses: Vec<BlueyResponse>,
    pub transcript_segment_count: u32,
}

/// Mirrors `SessionSearchQuery`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SessionSearchQuery {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mode_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub from: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub to: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub offset: Option<u32>,
}

File: src/lib/types/session.ts
/** Session domain types (mirrors `bluey_core::types::session`). */

import type { BlueyResponse } from "./response";

export type SessionStatus = "active" | "paused" | "completed";

export interface Session {
  id: string;
  modeId: string;
  startedAt: string;
  endedAt?: string;
  status: SessionStatus;
  title?: string;
  metadata?: Record<string, unknown>;
}

export type SessionEventType =
  | "session_started"
  | "session_paused"
  | "session_resumed"
  | "session_ended"
  | "question_detected"
  | "coding_problem_detected"
  | "objection_detected"
  | "decision_detected"
  | "action_item_detected"
  | "topic_change"
  | "important_statement"
  | "follow_up"
  | "screen_captured"
  | "response_prepared"
  | "response_generated"
  | "response_failed"
  | "document_attached"
  | "mode_changed"
  | "note_added"
  | "summary_generated"
  | "recording_imported";

export interface SessionEvent {
  id: string;
  sessionId: string;
  type: SessionEventType;
  /** Short human label shown in the timeline, e.g. "Question detected". */
  title: string;
  /** Optional detail (question text, decision text...). */
  detail?: string;
  /** Related entity ids for inspection (responseId, transcriptSegmentId, snapshotId...). */
  refs?: Record<string, string>;
  confidence?: number;
  createdAt: string;
}

export interface SessionNote {
  id: string;
  sessionId: string;
  content: string;
  createdAt: string;
  updatedAt: string;
}

export interface SessionSummary {
  id: string;
  sessionId: string;
  modeId: string;
  overview: string;
  topics: string[];
  questions: string[];
  answers: string[];
  decisions: string[];
  actionItems: string[];
  openItems: string[];
  improvements: string[];
  /** Mode-specific extra sections (e.g. lecture study guide). */
  sections?: Array<{ title: string; content: string }>;
  createdAt: string;
}

export interface SessionListItem {
  session: Session;
  modeName: string;
  eventCount: number;
  responseCount: number;
  transcriptSegmentCount: number;
  hasSummary: boolean;
  /** Search snippet when returned from search. */
  snippet?: string;
}

export interface SessionDetail {
  session: Session;
  events: SessionEvent[];
  notes: SessionNote[];
  summary?: SessionSummary;
  responses: BlueyResponse[];
  transcriptSegmentCount: number;
}

export interface SessionSearchQuery {
  text?: string;
  modeId?: string;
  from?: string;
  to?: string;
  limit?: number;
  offset?: number;
}
